cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java

```diff
@@ -0,0 +1,177 @@
+   * @param hBaseTableUtil Hbase table util
+   * @param maxRecordsToScanResolutionMap linked hash map of resolutions to perform migration
+   *                                      to the max records to scan for that resolution table
+   *                                      while performing migration
+   * @param executorService scheduled executor service
+   */
+  public DataMigrator(MetricDatasetFactory metricDatasetFactory, CConfiguration cConf, Configuration hConf,
+                      HBaseTableUtil hBaseTableUtil, LinkedHashMap<Integer, Integer> maxRecordsToScanResolutionMap,
+                      ScheduledExecutorService executorService) {
+    this.migrationTableUtility = new MigrationTableUtility(cConf, hConf, metricDatasetFactory, hBaseTableUtil);
+    this.maxRecordsToScanResolutionMap = maxRecordsToScanResolutionMap;
+    this.scheduledExecutorService = executorService;
+    this.cConf = cConf;
+  }
+
+  /**
+   * if the metrics data migration is already completed, we would just shutdown the executor service and return
+   * if not, we would schedule migration task at fixed rate schedule.
+   * when the migration is complete, the task would shutdown the executor service by itself
+   */
+  public void scheduleMigrationIfNecessary() {
+    if (!isMigrationComplete()) {
+      // we have 5 minute initial delay for availability of transaction/dataset service
+      // to be able to get the metrics table or if v3 tables are not already created they might take some time
+      // initially, we don't have to perform
+      scheduledExecutorService.scheduleAtFixedRate(new MigrationTask(), 5, 1, TimeUnit.MINUTES);
+      LOG.info("Scheduled metrics migration thread for resolution INT_MAX, 3600, 60 tables");
+    } else {
+      scheduledExecutorService.shutdown();
+    }
+  }
+
+  /**
+   * returns true when data migration is complete and
+   * all v2 metrics tables data has been migrated for configured resolutions ; false otherwise
+   * @return true if migration is complete; false otherwise
+   */
+  private boolean isMigrationComplete() {
+    Iterator<Integer> resolutions = maxRecordsToScanResolutionMap.keySet().iterator();
+
+    if (!resolutions.hasNext()) {
+      // all tables in migrationOrder map have been deleted, so migration is complete;
+      // useful when metrics.processor is running and completed migration
+      // and the run method removed all the entries from map when the tables were deleted
+      return true;
+    }
+
+    // when metrics.processor restarts and initializes DataMigrator,
+    // this is useful to decide whether to schedule migration thread or not
+    while (resolutions.hasNext()) {
+      int resolution = resolutions.next();
+      if (migrationTableUtility.v2MetricsTableExists(resolution)) {
+        try (MetricsTable v3MetricsTable = migrationTableUtility.getV3MetricsTable(resolution);
+             MetricsTable v2MetricsTable = migrationTableUtility.getV2MetricsTable(resolution)) {
```

Comment:
I guess the name of the method (get...Table) suggests that this will fail if there is an exception. It would be clearer if the method name indicates that it returns null if there is an exception. 

Also, is this true for all possible exceptions, that you want to ignore them and try again next cycle?